File: binary_tree/binary_tree.cpp
#include <iostream>
using namespace std;

class Node {
public:
    int data;
    Node* left;
    Node* right;
    
    Node(int data) {
        this->data = data;
        left = right = nullptr;
    }
};

class BinaryTree {
private:
    Node* root;

    Node* insertHelper(Node* root, Node* node) {
        if (root == nullptr) {
            return node;
        }
        if (node->data < root->data) {
            root->left = insertHelper(root->left, node);
        } else {
            root->right = insertHelper(root->right, node);
        }
        return root;
    }

    void displayHelper(Node* root) {
        if (root != nullptr) {
            displayHelper(root->left);
            cout << root->data << " ";
            displayHelper(root->right);
        }
    }

    bool searchHelper(Node* root, int data) {
        if (root == nullptr) return false;
        if (root->data == data) return true;
        if (data < root->data) return searchHelper(root->left, data);
        return searchHelper(root->right, data);
    }

    Node* removeHelper(Node* root, int data) {
        if (root == nullptr) {
            return root;
        } else if (data < root->data) {
            root->left = removeHelper(root->left, data);
        } else if (data > root->data) {
            root->right = removeHelper(root->right, data);
        } else {
            if (root->left == nullptr && root->right == nullptr) {
                delete root;
                return nullptr;
            } else if (root->right != nullptr) {
                root->data = successor(root);
                root->right = removeHelper(root->right, root->data);
            } else {
                root->data = predecessor(root);
                root->left = removeHelper(root->left, root->data);
            }
        }
        return root;
    }

    int successor(Node* root) {
        root = root->right;
        while (root->left != nullptr) {
            root = root->left;
        }
        return root->data;
    }

    int predecessor(Node* root) {
        root = root->left;
        while (root->right != nullptr) {
            root = root->right;
        }
        return root->data;
    }
    
public:
    BinaryTree() { root = nullptr; }

    void insertNode(int data) {
        root = insertHelper(root, new Node(data));
    }

    void display() {
        displayHelper(root);
        cout << endl;
    }

    bool search(int data) {
        return searchHelper(root, data);
    }

    void remove(int data) {
        if (search(data)) {
            root = removeHelper(root, data);
        } else {
            cout << "Couldn't find the data." << endl;
        }
    }
};

int main() {
    BinaryTree tree;
    tree.insertNode(5);
    tree.insertNode(1);
    tree.insertNode(9);
    tree.insertNode(2);
    tree.insertNode(7);
    tree.insertNode(3);
    tree.insertNode(6);
    tree.insertNode(4);
    tree.insertNode(8);

    tree.remove(4);
    tree.display();
    cout << tree.search(4) << endl;

    return 0;
}

File: binary_tree/binaryTree.py
class Node :
    def __init__(self,data):
        self.data = data
        self.left = None
        self.right = None

class BinaryTree :
    def __init__(self):
        self.root = None   
    
    def insert(self,data):
        self.root = self.insertHelper(self.root,data)
    
    def insertHelper(self,root,data):
        if root is None :
            return Node(data)
        if data < root.data :
            root.left = self.insertHelper(root.left,data)
        else:
            root.right = self.insertHelper(root.right,data)

        return root

    def display(self):
        self.displayHelper(self.root)
        print()

    def displayHelper(self,root):
        if root :
            self.displayHelper(root.left)
            print(root.data,end=" ")
            self.displayHelper(root.right)
    def search(self,data):
        return self.searchHelper(self.root,data)

    def searchHelper(self,root,data):
        if root is None :
            return False
        if root.data == data :
            return True
        if data < root.data:
            return self.searchHelper(root.left, data)
        else:
            return self.searchHelper(root.right, data)
    
    def remove(self,data):
        if self.search(data):
            self.root = self.removeHelper(self.root,data)
        else :
            print("Couldn't find the data.")

    def removeHelper(self,root,data):
        if root is None :
            return root
        if data < root.data :
            root.left = self.removeHelper(root.left,data)
        elif data > root.data :
            root.right = self.removeHelper(root.right,data)
        else :
            if root.left is None and root.right is None :
                return None
            elif root.right is not None :
                root.data = self.successor(root)
                root.right = self.removeHelper(root.right,root.data)
            else :
                root.data = self.predecessor(root)
                root.left = self.removeHelper(root.left,root.data)
        return root
    
    def successor(self,root) :
        root = root.right
        while(root.left is not None) :
            root = root.left
        return root.data
    
    def predecessor(self,root):
        root = root.left
        while root.right is not None :
            root = root.right
        return root.data
    
if __name__ == "__main__" :
    tree = BinaryTree()
    for value in [5,1,9,2,7,3,6,4,8] :
        tree.insert(value)
    
    tree.remove(4)
    tree.display()
    print(tree.search(4))

    